frame/composable-traits/src/staking_rewards.rs

```diff
@@ -0,0 +1,169 @@
+
+pub enum PositionState {
+	/// The position is not being rewarded yet and waiting for the next epoch.
+	Pending,
+	/// The position is currently locked and being rewarded.
+	LockedRewarding,
+	/// The position expired but still being rewarded.
+	Expired,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
+pub enum ClaimStrategy {
+	/// Basic, canonical claiming.
+	Canonical,
+	/// Force claiming the reward by restaking if the position expired.
+	RestakeIfExpired,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
+pub struct StakingConfig<AccountId, DurationPresets, Rewards> {
+	/// The possible locking duration.
+	pub duration_presets: DurationPresets,
+	/// The assets we can reward stakers with.
+	pub rewards: Rewards,
+	/// The penalty applied if a staker unstake before the end date.
+	pub early_unstake_penalty: Perbill,
+	/// The beneficiary of the penalty.
+	pub penalty_beneficiary: AccountId,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Encode, Decode, TypeInfo)]
+pub struct StakingNFT<AssetId, Balance, CollectedRewards> {
+	/// The staked asset.
+	pub asset: AssetId,
+	/// The stake this NFT was minted for.
+	pub stake: Balance,
+	/// The date at which this NFT was minted.
+	pub lock_date: Timestamp,
+	/// The duration for which this NFT stake was locked.
+	pub lock_duration: DurationSeconds,
+	/// The collected rewards counters at which this NFT was minted, used to compute the rewards.
+	pub collected_rewards: CollectedRewards,
+	/// The reward multiplier.
+	pub reward_multiplier: Perbill,
+}
+
+impl<AssetId, Balance: AtLeast32BitUnsigned + Copy, CollectedRewards>
+	StakingNFT<AssetId, Balance, CollectedRewards>
+{
+	pub fn penalize(&self, penalty: Perbill) -> Result<(Balance, Balance), DispatchError> {
+		let penalty_amount = penalty.mul_floor(self.stake);
+		let penalized_amount = self.stake.safe_sub(&penalty_amount)?;
+		Ok((penalized_amount, penalty_amount))
+	}
+
+	pub fn shares(&self) -> u128 {
+		self.reward_multiplier.mul_floor(self.stake.saturated_into::<u128>())
+	}
+
+	pub fn state(&self, now: Timestamp, epoch_start: Timestamp) -> PositionState {
+		if self.lock_date.saturating_add(self.lock_duration) < now {
+			PositionState::Expired
+		} else if self.lock_date < epoch_start {
+			PositionState::LockedRewarding
+		} else {
+			PositionState::Pending
+		}
+	}
+}
+
+impl<AssetId, Balance, CollectedRewards> Get<NFTClass>
+	for StakingNFT<AssetId, Balance, CollectedRewards>
+{
+	fn get() -> NFTClass {
+		NFTClass::STAKING
+	}
+}
+
+impl<AssetId, Balance, CollectedRewards> Get<NFTVersion>
+	for StakingNFT<AssetId, Balance, CollectedRewards>
+{
+	fn get() -> NFTVersion {
+		NFTVersion::VERSION_1
+	}
+}
+
+/// Interface for protocol staking.
+pub trait Staking {
+	type AccountId;
+	type AssetId;
+	type Balance;
+	type InstanceId;
+
+	/// Stake an amount of protocol asset. A new NFT representing the user position will be
```

Comment:
I wonder if this is sufficient documentation to explain the mechanism at work here (our slack chat helped a lot for me :) ). May be a link to further documentation might help? 
eg: https://app.clickup.com/20465559/docs/kghwq-63021 or a readme in the new staking-rewards pallet?